cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableAdmin.java

```diff
@@ -47,42 +49,48 @@
 
   private final DatasetDefinition tableDefinition;
   private final DatasetSpecification entityTableSpec;
+  private final StructuredTableRegistry registry;
 
   @Inject
-  public NoSqlStructuredTableAdmin(@Named(Constants.Dataset.TABLE_TYPE) DatasetDefinition tableDefinition) {
+  public NoSqlStructuredTableAdmin(@Named(Constants.Dataset.TABLE_TYPE) DatasetDefinition tableDefinition,
+                                   StructuredTableRegistry registry) {
     this.tableDefinition = tableDefinition;
     entityTableSpec = tableDefinition.configure(ENTITY_TABLE_NAME, DatasetProperties.EMPTY);
+    this.registry = registry;
   }
 
   @Override
-  public void create(StructuredTableSpecification spec) throws IOException {
+  public void create(StructuredTableSpecification spec) throws IOException, TableAlreadyExistsException {
     LOG.info("Creating table {} in namespace {}", spec.getTableId().getName(), NamespaceId.SYSTEM);
     DatasetAdmin admin = tableDefinition.getAdmin(SYSTEM_CONTEXT, entityTableSpec, null);
     if (!admin.exists()) {
       LOG.info("Creating dataset table {} in namespace {}", entityTableSpec.getName(), NamespaceId.SYSTEM);
       admin.create();
     }
-    StructuredTableSpecificationRegistry.registerSpecification(spec);
+    registry.registerSpecification(spec);
   }
 
   @Override
   public StructuredTableSpecification getSpecification(StructuredTableId tableId) {
-    return StructuredTableSpecificationRegistry.getSpecification(tableId);
+    return registry.getSpecification(tableId);
   }
 
   @Override
   public void drop(StructuredTableId tableId) throws IOException {
     LOG.info("Dropping table {} in namespace {}", tableId.getName(), NamespaceId.SYSTEM);
-    StructuredTableSpecificationRegistry.removeSpecification(tableId);
-    if (StructuredTableSpecificationRegistry.isEmpty()) {
+    registry.removeSpecification(tableId);
+    if (registry.isEmpty()) {
       DatasetAdmin admin = tableDefinition.getAdmin(SYSTEM_CONTEXT, entityTableSpec, null);
       LOG.info("Dropping dataset table {} in namespace {}", entityTableSpec.getName(), NamespaceId.SYSTEM);
       admin.drop();
     }
   }
 
-  <T> T getEntityTable() throws IOException {
-    //noinspection unchecked
-    return (T) tableDefinition.getDataset(SYSTEM_CONTEXT, entityTableSpec, Collections.emptyMap(), null);
+  <T> T getEntityTable(String name) throws IOException {
+    if (NoSqlStructuredTableAdmin.ENTITY_TABLE_NAME.equals(name)) {
```

Comment:
If the name has to be `ENTITY_TABLE_NAME`, what's the point of having it pass in from the parameter? Why not just the constant directly?

Reply:
Moved the check to `TableDatasetSupplier` instance